File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void generate(vector<int>& cand, int tar, int i, vector<vector<int>>& ans, vector<int> arr) {
        if(tar==0) {
            ans.push_back(arr);
            return;
        }
        if (i == cand.size() || tar < 0) {
            return;
        }
        arr.push_back(cand[i]);
        generate(cand, tar-cand[i], i+1, ans, arr);
        arr.pop_back();
        int n = i+1;
        while(n < cand.size() && cand[n] == cand[i]) n++;
        generate(cand, tar, n, ans, arr);
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> comb;
        generate(candidates, target, 0, ans, comb);
        return ans;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    void ps(int[] arr, int i, List<Integer> ans, List<List<Integer>> allSets, int t) {
        if(t == 0) {
            allSets.add(new ArrayList<>(ans));
            return;
        }
        if(i == arr.length || t<0) {
            return;
        }
        ans.add(arr[i]);
        ps(arr, i+1, ans, allSets, t-arr[i]);
        ans.remove(ans.size()-1);
        int idx = i+1;
        while(idx<arr.length && arr[idx] == arr[idx-1])idx++;
        ps(arr, idx, ans, allSets, t);
    }
    public List<List<Integer>> combinationSum2(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> allSets = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();
        ps(nums, 0, ans, allSets, target);
        return allSets;
    }
}